internal/querywatcher/query_watcher.go

```diff
@@ -121,35 +125,55 @@ func (w *QueryWatcher) watchKeys(ctx context.Context, watchChan <-chan dstore.Wa
 }
 
 // processWatchEvent processes a single watch event.
-func (w *QueryWatcher) processWatchEvent(event dstore.WatchEvent) {
-	w.WatchList.Range(func(key, value interface{}) bool {
-		query := key.(DSQLQuery)
-		clients := value.(*sync.Map)
-
-		if !regex.WildCardMatch(query.KeyRegex, event.Key) {
+func (w *QWatchManager) processWatchEvent(event dstore.WatchEvent) {
+	/*
+		Iterate over the watchlist to go through all fingerprints and the corresponding client connections to those fingerprints
+	*/
+	w.WatchList.Range(func(fingerPrint, value interface{}) bool {
+		/*
+			Fetch all the queries that match the fingerprint. If the fingerprint is not found in the cache, then it is likely that the query has been deleted. In that case, we can skip the rest of the logic
+		*/
+		queries, ok := w.FingerPrintCache.Get(fingerPrint.(string))
+		if !ok {
+			log.Warnf("Fingerprint not found in cacheStore: %s", fingerPrint)
 			return true
 		}
 
-		w.updateQueryCache(&query, event)
-		queryResult, err := w.runQuery(&query)
-		if err != nil {
-			log.Error(err)
+		clients := value.(*sync.Map)
+		/*
+			For all the clients that are watching this fingerprint, we need to find the queries they had requested to run, and then send them responses based on the queries they had fired.
+		*/
+		clients.Range(func(clientKey, _ interface{}) bool {
+			clientFd := clientKey.(int)
+
+			query := queries[clientFd]
+			if !regex.WildCardMatch(query.KeyRegex, event.Key) {
+				return true
+			}
+
+			w.updateQueryCache(fingerPrint.(string), event)
+			queryResult, err := w.runQuery(query)
+			if err != nil {
+				log.Error(err)
+				return true
+			}
+
+			w.notifyClient(query, clientFd, queryResult)
 			return true
-		}
+		})
```

Comment:
I'm a little confused about this looping mechanism here.

Shouldn't it be structured as follows:
```golang
w.WatchList.Range(
    // Looping over all queries.

    query := key.(DSQLQuery)
    clients := value.(*sync.Map)

    if !regex.WildCardMatch() {
        // continue if no match
    }

    // 1. If you reach here it means we are interested in this query.
    
    // 2. Fetch this query's fingerprint, and update the cache (you could maintain a local map keeping track of which fingerprints' cache we have updated, to avoid dual updates - though the operations are idempotent so it may not be necessary)

    // 3. Execute the query and fetch the result.
    queryResult, err := w.runQuery(&query)

    // 4. Notify clients
    w.notifyClients(&query, clients, queryResult)
)
```

Reply:
Won't it be more efficient if we find the DSQLQuery instances that has the regex match for the event key and then update the corresponding caches. 
Each DSQLQuery instance can hold the fingerprint info using which we can find the corresponding cache.

Reply:
Moved the Watchlist to be at query level and changed the logic based on that. Also, modified the DSQLQuery to hold the fingerprint.